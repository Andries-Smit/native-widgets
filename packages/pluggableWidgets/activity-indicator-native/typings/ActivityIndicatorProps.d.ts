/**
 * This file was generated from ActivityIndicator.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { CSSProperties } from "react";

export interface ActivityIndicatorProps<Style> {
    name: string;
    style: Style[];
}

export interface ActivityIndicatorPreviewProps {
    className: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
}
