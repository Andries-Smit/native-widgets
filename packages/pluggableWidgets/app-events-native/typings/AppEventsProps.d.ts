/**
 * This file was generated from AppEvents.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { CSSProperties } from "react";
import { ActionValue } from "mendix";

export type TimerTypeEnum = "once" | "interval";

export interface AppEventsProps<Style> {
    name: string;
    style: Style[];
    onLoadAction?: ActionValue;
    onUnloadAction?: ActionValue;
    onResumeAction?: ActionValue;
    onResumeTimeout: number;
    onOnlineAction?: ActionValue;
    onOnlineTimeout: number;
    onOfflineAction?: ActionValue;
    onOfflineTimeout: number;
    timerType: TimerTypeEnum;
    delayTime: number;
    onTimeoutAction?: ActionValue;
}

export interface AppEventsPreviewProps {
    className: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    onLoadAction: {} | null;
    onUnloadAction: {} | null;
    onResumeAction: {} | null;
    onResumeTimeout: number | null;
    onOnlineAction: {} | null;
    onOnlineTimeout: number | null;
    onOfflineAction: {} | null;
    onOfflineTimeout: number | null;
    timerType: TimerTypeEnum;
    delayTime: number | null;
    onTimeoutAction: {} | null;
}
