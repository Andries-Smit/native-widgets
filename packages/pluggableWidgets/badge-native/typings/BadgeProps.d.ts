/**
 * This file was generated from Badge.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { CSSProperties } from "react";
import { ActionValue, DynamicValue } from "mendix";

export interface BadgeProps<Style> {
    name: string;
    style: Style[];
    caption: DynamicValue<string>;
    onClick?: ActionValue;
}

export interface BadgePreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    renderMode: "design" | "xray" | "structure";
    translate: (text: string) => string;
    caption: string;
    onClick: {} | null;
}
