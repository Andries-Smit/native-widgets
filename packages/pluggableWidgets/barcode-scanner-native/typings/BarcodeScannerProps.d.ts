/**
 * This file was generated from BarcodeScanner.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix UI Content Team
 */
import { CSSProperties } from "react";
import { ActionValue, EditableValue } from "mendix";

export interface BarcodeScannerProps<Style> {
    name: string;
    style: Style[];
    barcode: EditableValue<string>;
    showMask: boolean;
    showAnimatedLine: boolean;
    onDetect?: ActionValue;
}

export interface BarcodeScannerPreviewProps {
    className: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    barcode: string;
    showMask: boolean;
    showAnimatedLine: boolean;
    onDetect: {} | null;
}
