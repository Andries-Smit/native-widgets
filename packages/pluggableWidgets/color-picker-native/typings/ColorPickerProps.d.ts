/**
 * This file was generated from ColorPicker.xml
 * WARNING: All changes made to this file will be overwritten
 * @author Mendix Widgets Framework Team
 */
import { CSSProperties } from "react";
import { ActionValue, EditableValue } from "mendix";

export type FormatEnum = "hex" | "hsl" | "hsv" | "rgb";

export interface ColorPickerProps<Style> {
    name: string;
    style: Style[];
    color: EditableValue<string>;
    format: FormatEnum;
    showPreview: boolean;
    showSaturation: boolean;
    showLightness: boolean;
    showAlpha: boolean;
    onChange?: ActionValue;
}

export interface ColorPickerPreviewProps {
    /**
     * @deprecated Deprecated since version 9.18.0. Please use class property instead.
     */
    className: string;
    class: string;
    style: string;
    styleObject?: CSSProperties;
    readOnly: boolean;
    renderMode: "design" | "xray" | "structure";
    translate: (text: string) => string;
    color: string;
    format: FormatEnum;
    showPreview: boolean;
    showSaturation: boolean;
    showLightness: boolean;
    showAlpha: boolean;
    onChange: {} | null;
}
