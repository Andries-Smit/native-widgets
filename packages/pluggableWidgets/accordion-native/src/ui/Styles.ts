import { Style } from "@mendix/piw-native-utils-internal";
import { ViewStyle, TextStyle } from "react-native";

export interface AccordionStyle extends Style {
    container: ViewStyle;
    group: AccordionGroupStyle;
}

export interface AccordionGroupStyle {
    container: ViewStyle;
    header: {
        container: ViewStyle;
        heading1: TextStyle;
        heading2: TextStyle;
        heading3: TextStyle;
        heading4: TextStyle;
        heading5: TextStyle;
        heading6: TextStyle;
        icon: AccordionIconStyle;
    };
    content: ViewStyle;
}

export interface AccordionIconStyle extends ViewStyle {
    size: number;
    color: string;
}

export const defaultAccordionStyle: AccordionStyle = {
    container: {
        flex: 1,
        backgroundColor: "#FFF",
        borderColor: "#CED0D3"
    },
    group: {
        container: {
            flex: 1,
            borderWidth: 1,
            borderColor: "#CED0D3"
        },
        header: {
            container: {
                flex: 1,
                flexDirection: "row",
                alignItems: "center",
                justifyContent: "space-between",
                paddingVertical: 16,
                paddingHorizontal: 16
            },
            heading1: {
                fontWeight: "600",
                fontSize: 40,
                lineHeight: 60
            },
            heading2: {
                fontWeight: "600",
                fontSize: 34,
                lineHeight: 51
            },
            heading3: {
                fontWeight: "600",
                fontSize: 28,
                lineHeight: 42
            },
            heading4: {
                fontWeight: "600",
                fontSize: 24,
                lineHeight: 36
            },
            heading5: {
                fontWeight: "600",
                fontSize: 20,
                lineHeight: 30
            },
            heading6: {
                fontWeight: "600",
                fontSize: 16,
                lineHeight: 24
            },
            icon: {
                height: 16,
                width: 16,
                size: 16,
                color: "#000"
            }
        },
        content: {
            paddingTop: 8,
            paddingBottom: 24,
            paddingHorizontal: 16
        }
    }
};
