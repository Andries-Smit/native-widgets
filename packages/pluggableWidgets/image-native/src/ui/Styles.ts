import { ViewStyle, ImageStyle } from "react-native";
import { Style } from "@mendix/piw-native-utils-internal";
import { GProps } from "react-native-svg";

export interface IconStyle {
    size?: number;
    color?: string;
}
export interface DefaultImageStyle extends Style {
    container: ViewStyle;
    image: ImageStyle | GProps | IconStyle;
    backdrop: ViewStyle;
}

export const defaultImageStyle: DefaultImageStyle = {
    container: {
        alignItems: "center",
        justifyContent: "center"
    },
    image: {
        flex: 1, // This is required to prevent blurry images on Android
        minHeight: 8,
        minWidth: 8,
        maxWidth: "100%",
        maxHeight: "100%"
    },
    backdrop: {
        width: "100%",
        height: "100%",
        justifyContent: "center",
        alignItems: "center",
        backgroundColor: `rgba(0,0,0,0.8)`
    }
};
