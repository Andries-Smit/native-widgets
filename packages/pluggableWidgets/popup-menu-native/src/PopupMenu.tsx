import { ComponentType, createElement, ReactElement, Fragment, useState } from "react";
import { PopupMenuProps } from "../typings/PopupMenuProps";
import { PopupMenuStyle } from "./ui/Styles";
import { executeAction } from "@mendix/piw-utils-internal";
import {
    Platform,
    StyleSheet,
    TouchableHighlight,
    TouchableHighlightProps,
    TouchableNativeFeedback,
    TouchableNativeFeedbackProps,
    TouchableOpacity,
    View
} from "react-native";
import { ActionValue } from "mendix";
import { Menu, MenuDivider, MenuItem } from "react-native-material-menu";

const TouchableItem: ComponentType<TouchableNativeFeedbackProps | TouchableHighlightProps> =
    Platform.OS === "android" ? TouchableNativeFeedback : TouchableHighlight;

const TouchableButton: ComponentType<TouchableNativeFeedbackProps | TouchableOpacity> =
    Platform.OS === "android" ? TouchableNativeFeedback : TouchableOpacity;

export function PopupMenu(props: PopupMenuProps<PopupMenuStyle>): ReactElement {
    const styles = StyleSheet.flatten(props.style);
    const [visible, setVisible] = useState(false);

    const showMenu = () => setVisible(true);

    const handlePress = (action?: ActionValue) => {
        setVisible(false);
        // Set timeout needed since modal closes the alerts which might be shown in action
        setTimeout(() => {
            executeAction(action);
        }, 500);
    };

    let menuOptions: ReactElement[];
    if (props.renderMode === "basic") {
        menuOptions = props.basicItems.map((item, index) => {
            const itemStyle = styles.basic?.itemStyle && styles.basic?.itemStyle[item.styleClass];
            return item.itemType === "divider" ? (
                <MenuDivider key={index} color={styles.basic?.dividerColor} />
            ) : (
                <MenuItem
                    key={index}
                    onPress={() => handlePress(item.action)}
                    textStyle={itemStyle}
                    style={({ pressed }) => ({
                        ...styles.basic?.container,
                        ...(pressed ? { backgroundColor: styles.basic?.itemStyle?.rippleColor } : {})
                    })}
                    {...(Platform.OS === "android" && styles.basic?.itemStyle?.rippleColor
                        ? { android_ripple: { color: styles.basic.itemStyle.rippleColor, borderless: false } }
                        : undefined)}
                    testID={`${props.name}_basic-item`}
                >
                    {item.caption}
                </MenuItem>
            );
        });
    } else {
        menuOptions = props.customItems.map((item, index) => (
            <TouchableItem
                key={index}
                style={styles.custom?.container}
                onPress={() => handlePress(item.action)}
                {...getRippleColor(styles.custom?.itemStyle?.rippleColor)}
                testID={`${props.name}_custom-item`}
            >
                {Platform.OS === "android" ? (
                    <View style={styles.custom?.container}>{item.content}</View>
                ) : (
                    <Fragment>{item.content}</Fragment>
                )}
            </TouchableItem>
        ));
    }

    return (
        <Menu
            visible={visible}
            animationDuration={150}
            style={styles?.container as any}
            anchor={
                <TouchableButton onPress={showMenu} testID={`${props.name}_trigger`}>
                    <View pointerEvents="box-only" style={styles.buttonContainer}>
                        {props.menuTriggerer}
                    </View>
                </TouchableButton>
            }
        >
            <View style={{ overflow: "hidden", borderRadius: styles.container?.borderRadius }}>{menuOptions}</View>
        </Menu>
    );
}

function getRippleColor(color: string | undefined) {
    if (color) {
        return Platform.OS === "android"
            ? color && { background: TouchableNativeFeedback.Ripple(color, false) }
            : { underlayColor: color };
    }
    return undefined;
}
