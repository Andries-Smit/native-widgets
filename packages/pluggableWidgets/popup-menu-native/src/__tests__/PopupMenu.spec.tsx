import { PopupMenuProps } from "../../typings/PopupMenuProps";
import { PopupMenuStyle } from "../ui/Styles";
import { Text, View } from "react-native";
import { createElement } from "react";
import { actionValue } from "@mendix/piw-utils-internal";
import { fireEvent, render, within } from "@testing-library/react-native";
import { PopupMenu } from "../PopupMenu";
import { MenuDivider } from "react-native-material-menu";
import { ReactTestInstance } from "react-test-renderer";

let dummyActionValue: any;
let defaultProps: PopupMenuProps<PopupMenuStyle>;
jest.useFakeTimers();

let basicItemTestId: string;
let customItemTestId: string;

jest.mock("react-native/Libraries/Modal/Modal", () => (props: any) => {
    const MockedModal = jest.requireActual("react-native").View;
    return <MockedModal {...props} testID="modal" />;
});

describe("Popup menu", () => {
    beforeEach(() => {
        dummyActionValue = actionValue();
        defaultProps = {
            renderMode: "basic",
            name: "popup-menu",
            style: [],
            menuTriggerer: <Text>Menu Triggerer</Text>,
            basicItems: [
                { itemType: "item", action: dummyActionValue, caption: "yolo", styleClass: "defaultStyle" },
                { itemType: "divider", styleClass: "defaultStyle", caption: "" }
            ],
            customItems: [{ content: <Text>Yolo</Text>, action: dummyActionValue }]
        };

        basicItemTestId = `${defaultProps.name}_basic-item`;
        customItemTestId = `${defaultProps.name}_custom-item`;
    });

    it("renders", () => {
        const component = render(<PopupMenu {...defaultProps} />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    it("renders custom items", () => {
        const component = render(<PopupMenu {...defaultProps} renderMode="custom" />);

        expect(component.toJSON()).toMatchSnapshot();
    });

    describe("in basic mode", () => {
        it("renders only basic items", async () => {
            const component = render(<PopupMenu {...defaultProps} />);

            const basicItems = component.queryAllByTestId(basicItemTestId);
            const dividers = component.UNSAFE_queryAllByType(MenuDivider);
            const customItems = component.queryAllByTestId(customItemTestId);
            expect(basicItems).toHaveLength(1);
            expect(within(basicItems[0]).getByText("yolo")).not.toBeNull();
            expect(dividers).toHaveLength(1);
            expect(customItems).toHaveLength(0);
        });

        it("triggers action", () => {
            const component = render(<PopupMenu {...defaultProps} />);
            const basicItem = component.container.findByType(MenuItem);
            expect(basicItem).toBeDefined();
            expect(basicItem.props.testID).toBe(basicItemTestId);

            fireEvent.press(basicItem);

            jest.advanceTimersByTime(501);

            expect(dummyActionValue.execute).toHaveBeenCalled();
        });

        it("renders with custom styles", () => {
            const customStyle = [
                {
                    basic: {
                        itemStyle: {
                            defaultStyle: {
                                color: "green"
                            }
                        }
                    }
                }
            ];
            const component = render(<PopupMenu {...defaultProps} style={customStyle} />);
            expect(component.getByTestId(basicItemTestId).findByType(Text).props.style).toEqual(
                expect.arrayContaining([{ color: "green" }])
            );
        });
    });

    describe("in custom mode", () => {
        beforeEach(() => {
            defaultProps.renderMode = "custom";
        });

        it("renders only custom items", () => {
            const component = render(<PopupMenu {...defaultProps} />);

            const basicItems = component.queryAllByTestId(basicItemTestId);
            const dividers = component.UNSAFE_queryAllByType(MenuDivider);
            const customItems = component.queryAllByTestId(customItemTestId);
            expect(basicItems).toHaveLength(0);
            expect(dividers).toHaveLength(0);
            expect(customItems).toHaveLength(1);
            expect(within(customItems[0]).getByText("Yolo")).not.toBeNull();
        });

        it("triggers action", () => {
            const component = render(<PopupMenu {...defaultProps} />);
            fireEvent.press(component.getByTestId(customItemTestId));

            jest.advanceTimersByTime(501);

            expect(dummyActionValue.execute).toHaveBeenCalled();
        });

        it("renders with custom styles", () => {
            const customStyle = [
                {
                    container: {
                        backgroundColor: "yellow"
                    }
                }
            ];
            const component = render(<PopupMenu {...defaultProps} style={customStyle} />);
            const modal = component.getByTestId("modal");
            const firstView = within(modal).UNSAFE_getByType(View);
            const secondView = within(firstView.children[0] as ReactTestInstance).UNSAFE_getByType(View);
            expect(secondView.props.style.backgroundColor).toEqual("yellow");
            expect(within(secondView).getByText("Yolo")).not.toBeNull();
        });
    });
});
