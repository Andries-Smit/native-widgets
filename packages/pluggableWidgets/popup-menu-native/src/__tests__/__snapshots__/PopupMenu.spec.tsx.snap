// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup menu renders 1`] = `
<View
  collapsable={false}
>
  <View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        {
          "attribute": "selectableItemBackground",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      pointerEvents="box-only"
      testID="popup-menu_trigger"
    >
      <Text>
        Menu Triggerer
      </Text>
    </View>
  </View>
  <Modal
    hardwareAccelerated={false}
    onDismiss={[Function]}
    onRequestClose={[Function]}
    supportedOrientations={
      [
        "portrait",
        "portrait-upside-down",
        "landscape",
        "landscape-left",
        "landscape-right",
      ]
    }
    transparent={true}
    visible={false}
  />
</View>
`;
